### PR TITLE
cp-1605 : added jitter and maxDuration between retries

### PR DESCRIPTION
**Pull request by dustinlessard-wf**
### CODE REVIEW

This is a rebase of https://github.com/Workiva/w_transport/pull/155
For consideration:

#### PROBLEM
We don't currently allow jitter with our exponential backoff strategy.

#### SOLUTION
We now allow the optional inclusion of jitter and a maximum retry duration + default maximum retry duration.

Inspired by https://www.awsarchitectureblog.com/2015/03/backoff.html

**Testing Suggestions:**
run unit-tests
run integration tests

**Required:**
@evanweible-wf @maxwellpeterson-wf @trentgrover-wf @sebastianmalysa-wf 

**FYI:**
@jayudey-wf

### CODE REVIEW

File: lib/src/http/common/backoff.dart [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'dart:math';
+import 'package:w_transport/src/http/auto_retry.dart';
+
+class Backoff {
+  static const int defaultExponentialMultiplier = 2;
```
Comment (evanweible-wf) [inline lib/src/http/common/backoff.dart]: Would this ever not be 2? The `default` part seems misleading if you can't override it

File: lib/src/http/common/backoff.dart [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+  static Duration _createFixedBackOff(RequestAutoRetry autoRetry) {
+    Duration backOff;
+
+    if (autoRetry.backOff.withJitter == true) {
+      Random random = new Random();
+      backOff = new Duration(
+          milliseconds: autoRetry.backOff.interval.inMilliseconds ~/ 2 +
+              random.nextInt(
+                  (autoRetry.backOff.interval.inMilliseconds).toInt()));
```
Comment (evanweible-wf) [inline lib/src/http/common/backoff.dart]: These parens seem unnecessary

File: lib/src/http/common/backoff.dart [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    } else {
+      backOff = autoRetry.backOff.interval;
+    }
+
+    return backOff;
+  }
+
+  /// Calculate the backoff duration based on [RequestAutoRetry] configuration.
```
Comment (evanweible-wf) [inline lib/src/http/common/backoff.dart]: This comment could be expanded to note that `null` will be returned if backoff is not applicable

File: lib/src/http/common/request.dart [hunk 4]
```diff
@@ -31,6 +31,8 @@ import 'package:w_transport/src/http/request_exception.dart';
 import 'package:w_transport/src/http/request_progress.dart';
 import 'package:w_transport/src/http/requests.dart';
 import 'package:w_transport/src/http/response.dart';
+import 'dart:math';
```
Comment (evanweible-wf) [inline lib/src/http/common/request.dart]: `dart:` import should go above, and the `src/http/common/` import below should go a little higher up to be alphabetized

File: lib/src/http/common/request.dart [hunk 5]
```diff
@@ -45,6 +47,8 @@ abstract class CommonRequest extends Object
     autoRetry = new RequestAutoRetry(this);
   }
 
+  static const int defaultExponentialMultiplier = 2;
```
Comment (evanweible-wf) [inline lib/src/http/common/request.dart]: Is this no longer being used?

File: pubspec.yaml [hunk 6]
```diff
@@ -28,7 +28,7 @@ dev_dependencies:
   args: "^0.13.0"
   browser: "^0.10.0+2"
   coverage: "^0.7.4"
-  dart_dev: "^1.0.0"
+  dart_dev: "^1.1.2"
```
Comment (evanweible-wf) [inline pubspec.yaml]: Just curious, did you hit an issue that required this change?

File: lib/src/http/auto_retry.dart [hunk 7]
```diff
@@ -157,31 +157,48 @@ class RequestAutoRetry extends AutoRetryConfig {
...
 
   /// Construct a null back-off representation, meaning no delay between retry
   /// attempts.
   const RetryBackOff.none()
-      : duration = null,
-        method = RetryBackOffMethod.none;
+      : interval = null,
+        method = RetryBackOffMethod.none,
+        withJitter = false,
+        maxInterval = null;
```
Comment (evanweible-wf) [inline lib/src/http/auto_retry.dart]: This `maxInterval = null` assignment is unnecessary, same with the one above for `.fixed()`

File: lib/src/http/common/backoff.dart (hunk 1 above)
Comment (dustinlessard-wf) [inline lib/src/http/common/backoff.dart]: Agreed, renamed to `base`.

File: lib/src/http/common/backoff.dart (hunk 2 above)
Comment (dustinlessard-wf) [inline lib/src/http/common/backoff.dart]: removed

File: lib/src/http/common/backoff.dart (hunk 3 above)
Comment (dustinlessard-wf) [inline lib/src/http/common/backoff.dart]: Done

File: lib/src/http/common/request.dart (hunk 4 above)
Comment (dustinlessard-wf) [inline lib/src/http/common/request.dart]: Addressed, removed as the import wasn't being used.

File: lib/src/http/common/request.dart (hunk 5 above)
Comment (dustinlessard-wf) [inline lib/src/http/common/request.dart]: Removed.

File: pubspec.yaml (hunk 6 above)
Comment (dustinlessard-wf) [inline pubspec.yaml]: No, just noticed there was a newer version.

File: lib/src/http/auto_retry.dart (hunk 7 above)
Comment (dustinlessard-wf) [inline lib/src/http/auto_retry.dart]: It ends up being an analyzer error if we remove this.
screen shot 2016-06-22 at 9 29 06 am

Comment (evanweible-wf) [inline lib/src/http/auto_retry.dart]: Ah that's right, forgot about that 👍

**Comment by evanweible-wf** [comment]
@dustinlessard-wf something's borked, dartfmt failed to run because it couldn't parse one of the files

**Comment by sebastianmalysa-wf** [comment]
+1

**Comment by evanweible-wf** [comment]
@dustinlessard-wf Looks good, could you squash the commits?

**Comment by evanweible-wf** [comment]
+1

File: test/unit/platform_independent_unit_test.dart [hunk 8]
```diff
@@ -63,6 +64,5 @@ void main() {
   mock_web_socket_test.main();
 
   ws_w_socket_exception_test.main();
-  ws_w_socket_subscription_test.main();
```
Comment (evanweible-wf) [inline test/unit/platform_independent_unit_test.dart]: @dustinlessard-wf Wait why was this removed?

Comment (dustinlessard-wf) [inline test/unit/platform_independent_unit_test.dart]: I have no idea.  One sec, I'll restore that.

**Comment by evanweible-wf** [comment]
+1

**Comment by maxwellpeterson-wf** [comment]
+1

**Comment by dustinlessard-wf** [comment]
Ready for you @jayudey-wf

**Comment by jayudey-wf** [comment]
QA Resource Approval: +10

- [ ] Testing instruction
- [ ] Dev +1's
- [ ] Dev/QA +10 with detail of what was tested
 - passing CI with tests exercising functionality
- [ ] Unit test created/updated
- [ ] All unit tests pass 

Merging into master.
